### ISSUE TITLE
Method search in Bio.Restriction.AllEnzymes

### ISSUE THREAD

**Issue by paxelito** [biopython/biopython#1113]
``` python

from Bio.Seq import Seq
from Bio.Alphabet.IUPAC import IUPACAmbiguousDNA
from Bio.Restriction import AllEnzymes, CommOnly
my_seq = Seq('ACCCGAATTCAAAACTGACTGATCGATCGTCGACTG', IUPACAmbiguousDNA())
print(AllEnzymes.search(my_seq))
Traceback (most recent call last):
  File "scripts/rest_enz_test.py", line 55, in <module>
    ful_search = AllEnzymes.search(my_seq)
AttributeError: 'set' object has no attribute 'search'
```

It works perfectly with CommOnly but not with AllEnzymes RestritionBatch 

Python 3.6.0
Biopython 1.68
Numpy 1.12.0 

THX in advance

**Comment by MarkusPiotrowski** [comment]
This seems to be a Pyton 2/3 problem.

Works fine for me under Phyton 2.7.11, Biopython 1.68

```python
>>> from Bio.Seq import Seq
>>> from Bio.Alphabet.IUPAC import IUPACAmbiguousDNA
>>> from Bio.Restriction import AllEnzymes, CommOnly
>>> my_seq = Seq('ACCCGAATTCAAAACTGACTGATCGATCGTCGACTG', IUPACAmbiguousDNA())
>>> print(AllEnzymes.search(my_seq))
{BspQI: [], SgrBI: [], MluCI: [6], AoxI: [], BsmFI: [], AluI: [], PleI: [], PshA
...
```
However, the same error on Python 3.5.1, Biopython 1.68:
```python
>>> from Bio.Seq import Seq
>>> from Bio.Alphabet.IUPAC import IUPACAmbiguousDNA
>>> from Bio.Restriction import AllEnzymes, CommOnly
>>> my_seq = Seq('ACCCGAATTCAAAACTGACTGATCGATCGTCGACTG', IUPACAmbiguousDNA())
>>> print(AllEnzymes.search(my_seq))
Traceback (most recent call last):
  File "<pyshell#4>", line 1, in <module>
    print(AllEnzymes.search(my_seq))
AttributeError: 'set' object has no attribute 'search'
```
The problem seems to be this line (2575) in ```Restriction.py``` 
```python
AllEnzymes = CommOnly | NonComm
```
because in Python 2.7 this gives the same type:
```python
>>>from Bio.Restriction import NonComm, CommOnly, AllEnzymes
>>> type(NonComm)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
>>> type(CommOnly)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
>>> type(AllEnzymes)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
```
but in Python 3 not:
```python
>>> type(NonComm)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
>>> type(CommOnly)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
>>> type(AllEnzymes)
<class 'set'>
```

**Comment by peterjc** [comment]
Good detective work @MarkusPiotrowski - this shows if nothing else our test coverage is lacking here.

It looks like this is just a subclass of ``set`` with no explicit ``__or__`` method:

```
class RestrictionBatch(set):
    ...
```

Curious.

**Comment by peterjc** [comment]
@paxelito's example can be turned into a unit test easily, which is great:

```
$ git diff
diff --git a/Tests/test_Restriction.py b/Tests/test_Restriction.py
index f455388b1..fe61dcbcb 100644
--- a/Tests/test_Restriction.py
+++ b/Tests/test_Restriction.py
@@ -6,7 +6,7 @@
 """Testing code for Restriction enzyme classes of Biopython.
 """
 
-from Bio.Restriction import Analysis, Restriction, RestrictionBatch
+from Bio.Restriction import Analysis, Restriction, RestrictionBatch, AllEnzymes
 from Bio.Restriction import Acc65I, Asp718I, EcoRI, EcoRV, KpnI, SmaI
 from Bio.Seq import Seq
 from Bio.Alphabet.IUPAC import IUPACAmbiguousDNA
@@ -258,6 +258,11 @@ class RestrictionBatches(unittest.TestCase):
         self.assertEqual((ana._boundaries(20, 1)[:2]), (1, 20))  # reverse order
         self.assertEqual((ana._boundaries(-1, 20)[:2]), (20, 33))  # fix negative start
 
+    def test_search(self):
+        my_seq = Seq('ACCCGAATTCAAAACTGACTGATCGATCGTCGACTG', IUPACAmbiguousDNA())
+        search = AllEnzymes.search(my_seq)
+        self.assertTrue(" MluCI: [6], " in str(search), str(search))
+
 
 if __name__ == "__main__":
     runner = unittest.TextTestRunner(verbosity=2)
```

**Comment by MarkusPiotrowski** [comment]
I guess the problem can be overcome by replacing the ```|``` operator with an explicit ```update()``` method:
```python
>>> my_AllEnzymes = CommOnly
>>> type(my_AllEnzymes)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
>>> my_AllEnzymes.update(NonComm)
>>> type(my_AllEnzymes)
<class 'Bio.Restriction.Restriction.RestrictionBatch'>
```
which works under Python 2 and 3:
```python
>>> my_AllEnzymes.search(my_seq)
{BstMAI: [], TauI: [], BstH2I: [], TseI: [], BstZI: [], PalAI: [], GlaI: [], Sat
...
```

**Comment by peterjc** [comment]
Good plan Markus, could you turn that into a pull request along with the new test case please?

**Comment by MarkusPiotrowski** [comment]
Yes, I do. Hopefully I will also find the problem in #1128

**Comment by MarkusPiotrowski** [comment]
This was solved with  PR #1195.

Closing now.
